### PR TITLE
Add timing ratio histograms

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR adds timing ratio histograms, as an alternative to sample-and-watermark histograms,
and switches API Priority and Fairness metrics to use the timing ratio histograms.

#### Which issue(s) this PR fixes:

This partially addresses #108272

#### Special notes for your reviewer:
This is part of the campaign to replace sample-and-watermark histograms.
I am breaking #109742 into pieces for easier reviewing.
This is most of the middle piece; PRs #110515 and #110516, both now merged, are the rest of the middle piece.
There are more PRs to come; #110164 is next in line, and there will also be PRs that replace repeated indexing with element caching (as seen in #109742 and #110154) and do other cleanup.

#### Does this PR introduce a user-facing change?

```release-note
Some apiserver metrics were changed, as follows.
- `priority_level_seat_count_samples` is replaced with `priority_level_seat_utilization`, which samples every nanosecond rather than every millisecond; the old metric conveyed utilization despite its name.
- `priority_level_seat_count_watermarks` is removed.
- `priority_level_request_count_samples` is replaced with `priority_level_request_utilization`, which samples every nanosecond rather than every millisecond; the old metric conveyed utilization despite its name.
- `priority_level_request_count_watermarks` is removed.
- `read_vs_write_request_count_samples` is replaced with `read_vs_write_current_requests`, which samples every nanosecond rather than every second; the new metric, like the old one, measures utilization when the max-in-flight filter is used and number of requests when the API Priority and Fairness filter is used.
- `read_vs_write_request_count_watermarks` is removed.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/sig api-machinery
/cc @wojtek-t 
/cc @tkashem 
@deads2k 
@lavalamp

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
Here are the current benchmarking results.

```shell
(base) mspreitz@mjs12 metrics % go test  -benchmem -run=^$ -bench Histogram .
goos: darwin
goarch: amd64
pkg: k8s.io/apiserver/pkg/util/flowcontrol/metrics
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkSampleAndWaterMarkHistogramsVecEltSimple-16       	 1000000	      1230 ns/op	       0 B/op	       0 allocs/op
BenchmarkSampleAndWaterMarkHistogramsVecEltSafeEarly-16    	  987891	      1230 ns/op	       0 B/op	       0 allocs/op
BenchmarkSampleAndWaterMarkHistogramsVecEltSafeLate-16     	  982856	      1200 ns/op	       0 B/op	       0 allocs/op
BenchmarkTimingRatioHistogram-16                           	13167859	        90.70 ns/op	       0 B/op	       0 allocs/op
BenchmarkTimingRatioHistogramVecEltSimple-16               	13016456	        89.85 ns/op	       0 B/op	       0 allocs/op
BenchmarkTimingRatioHistogramVecEltSafeEarly-16            	 7515996	       160.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkTimingRatioHistogramVecEltSafeLate-16             	13404110	        89.83 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	k8s.io/apiserver/pkg/util/flowcontrol/metrics	9.665s
```

**Comment by MikeSpreitzer** [comment]
The force-push to 44a4470191d is a rebase onto master.

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight_test.go [hunk 1]
```diff
@@ -101,6 +102,8 @@ func TestMaxInFlightNonMutating(t *testing.T) {
 	waitForCalls := true
 	waitForCallsMutex := sync.Mutex{}
 
+	fcmetrics.Register()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight_test.go:105]: Can you instead Register those in createMaxInflightServer?

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go [hunk 2]
```diff
@@ -186,6 +186,7 @@ func newApfHandlerWithFilter(t *testing.T, flowControlFilter utilflowcontrol.Int
 
 func TestApfSkipLongRunningRequest(t *testing.T) {
 	epmetrics.Register()
+	fcmetrics.Register()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:189]: Same here - can we instead register them in newApf* helpers?

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:189]: [Maybe also move this cleanup (together with the maxinflight_test.go file) to a separate cleanup PR.]

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go [hunk 3]
```diff
@@ -57,6 +57,7 @@ func (spg SampleAndWaterMarkPairVec) NewForLabelValuesSafe(initialWaitingDenomin
 	}
 }
 
+//nolint:golint,unused
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:60]: If it's unused, let's remove it...

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go [hunk 4]
```diff
@@ -120,3 +120,105 @@ func getHistogramCount(registry compbasemetrics.KubeRegistry, metricName string)
...
+	for i := 0; i < b.N; i++ {
+		delta := (i % 6) + 1
+		now = now.Add(time.Duration(delta) * time.Millisecond)
+		clk.SetTime(now)
+		th.Set(float64(x))
+		x = (x + i) % 60
+	}
+}
+
+func BenchmarkSampleAndWaterMarkHistogramsVecEltSafeLate(b *testing.B) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go:192]: What's the different between this benchmark and the first one?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+// The `TimingHistogramOpts.InitialValue` is the initial numerator.
+type TimingRatioHistogramOpts struct {
+	compbasemetrics.TimingHistogramOpts
+	InitialDenominator float64
+}
+
+// timingRatioHistogramInner implements the instrumentation aspect
+type timingRatioHistogramInner struct {
+	nowFunc  func() time.Time
+	delegate interface{ Set(float64) }
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:55]: if we need an interface, can we define that instead of inlining it here?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+	trh.Lock()
+	defer trh.Unlock()
+	numerator := trh.numerator + deltaNumerator
+	trh.numerator = numerator
+	ratio := numerator / trh.denominator
+	trh.delegate.Set(ratio)
+}
+
+func (trh *timingRatioHistogramInner) Sub(deltaNumerator float64) {
+	trh.Lock()
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:103]: Can you just call trh.Add(-deltaNumerator) instead?

Same in Inc/Dec mehods.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+		return tro
+	}
+	return &timingRatioHistogramInner{
+		delegate:    &timingHistogramVecElt{delegate: v.delegate, labelValues: labelValues},
+		numerator:   initialNumerator,
+		denominator: initialDenominator,
+	}
+}
+
+type timingHistogramVecElt struct {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:226]: What is timingHistogramVecElt?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go [hunk 8]
```diff
@@ -120,3 +120,105 @@ func getHistogramCount(registry compbasemetrics.KubeRegistry, metricName string)
 	}
 	return 0, errMetricNotFound
 }
+
+func BenchmarkSampleAndWaterMarkHistogramsVecEltSimple(b *testing.B) {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go:124]: FWIW - benchmarks can also go to a separate PR.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go (hunk 3 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go:60]: Yes, that's coming later in a cleanup PR.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go:192]: Oh, right, the last edit made the first the same as the last.  I will remove the first.  It would be the parallel to BenchmarkTimingRatioHistogramVecEltSimple but can not exist because the sample-and-watermark histograms have no `NewForLabelValuesChecked`.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:55]: We could.  It would only be used internally to this one file, so I saw little value in defining it.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go (hunk 8 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go:124]: They _could_ but I chose to include them here because the point is to show that the timing ratio histograms added in this PR are much more efficient than the sample-and-watermark histograms.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go (hunk 4 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go:158]: done in the latest revision

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 5 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: Since you asked for it, I made it.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 6 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: done in the latest revision

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 7 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:211]: I added a comment explaining it.  It is in service of NewForLabelValuesSafe, whose comment I noticed was buggy, so I fixed that too.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go (hunk 8 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go]: The git commit message includes the benchmark results.

**Comment by MikeSpreitzer** [comment]
The force-push to 897a72ad1ed responds to the review above.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+	tro, err := v.NewForLabelValuesChecked(initialNumerator, initialDenominator, labelValues)
+	if err == nil {
+		return tro
+	}
+	if !compbasemetrics.ErrIsNotRegistered(err) {
+		klog.ErrorS(err, "Failed to extract TimingRatioHistogramVec member, using noop instead", "vectorname", v.FQName(), "labelValues", labelValues)
+		return tro
+	}
+	return &timingRatioHistogramInner{
+		delegate:    &timingHistogramVecElt{delegate: v.delegate, labelValues: labelValues},
```
[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:204]: And building based on this comment further:
- then you don't need `settable` interface at all, because it would always be of type GaugeMetric

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+func (noopRatioed) Dec()                   {}
+func (noopRatioed) SetToCurrentTime()      {}
+func (noopRatioed) SetDenominator(float64) {}
+
+func (v *TimingRatioHistogramVec) Reset() {
+	v.delegate.Reset()
+}
+
+type TimingRatioHistogramPairVec struct {
+	urVec *TimingRatioHistogramVec
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:235]: What is "urVec"?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 11]
```diff
@@ -0,0 +1,272 @@
...
+}
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:236]: A more generic comment is - why do we even need this type?

The more types we have the harder it is to follow when to use what, I don't see a need for this type.

Why can't we just instead create a simple helper functions for the TiminingRatioHistogramVec type, like
```
// better name needed desired
func (v *TimingRatioHistogramVec) WaitinExecutingGaugePair(nominator, denominator labels) {
}
```

**Review by wojtek-t — COMMENTED** [review]
Sorry for delay in review.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:204]: At this point `v.delegate.WithLabelValues(labelValues)` is guaranteed to return a permanent no-op, and the whole point here is to return something that is _not_ a permanent no-op.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 11 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:236]: It is a convenience.  See #110515 for its removal; not quite as small a thing as I suspect you were thinking.

File: staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight_test.go (hunk 1 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight_test.go:105]: Done in #110516

File: staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go (hunk 2 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:189]: #110516

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go (hunk 3 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+// The context is no-op at the current level of development.
+func (trh *timingRatioHistogramInner) WithContext(ctx context.Context) RatioedGauge {
+	return trh
+}
+
+// TimingRatioHistogramVec is a collection of TimingRatioHistograms that differ
+// only in label values.
+// `*TimingRatioHistogramVec` implements both Registerable and RatioedGaugeVec.
+type TimingRatioHistogramVec struct {
+	compbasemetrics.Registerable                                // promote only the Registerable methods
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: I added a comment explaining it.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 7 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: Revised to spell it out.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 10 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: gone

**Comment by MikeSpreitzer** [comment]
The recent force-pushes respond to the latest review.

**Comment by MikeSpreitzer** [comment]
The force-push to 6ae153140f4 is a rebase onto the `master` branch.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+// timingRatioHistogramInner implements the instrumentation aspect
+type timingRatioHistogramInner struct {
+	nowFunc  func() time.Time
+	delegate settable
+	sync.Mutex
+	// access only with mutex locked
+	numerator, denominator float64
+}
+
+type settable interface {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:61]: I still don't understand why you need to introduce this interface (yet another one).

You create timingRatioHistogramInner in two places:
- in one delegate is TimingHistogram
- in the other it's GaugeMetric 

Both are implementing the GaugeOps interface, so why not just make a delegate be of type
`GaugeOps` instead

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram_test.go [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+	toRegister := vec.metrics()
+	registry := compbasemetrics.NewKubeRegistry()
+	for _, reg := range toRegister {
+		registry.MustRegister(reg)
+	}
+	tro := vec.NewForLabelValuesSafe(0, 1, []string{"avalue"})
+	exerciseTimingRatioHistogram(t, samplesHistName, t0, clk, registry, tro)
+}
+
+/* exerciseTimingRatioHistogram does a rough behavioral test of a
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram_test.go:96]: nit: let's use `// in each line` comment style for consistency with the whole codebase

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram_test.go:96]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go (hunk 13 above)
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:61]: No, there is a third place, and it does not use a GaugeMetric: https://github.com/MikeSpreitzer/kubernetes/blob/add-timing-ratio-histograms/staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go#L211

Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:61]: OK - forgot about this. But I still think we can make it easier.
My proposal is the following:

1. we change delegate to be of type `func() GaugeOps`
1. In the places I mentioned above we simply set it to `func() { return <current value of delegate> }`
1. In the place you pointed above, we change it to:
   `func() GaugeOps { return tve.delegate.WithLabelValues(tve.labelValues...) }`

This allows us to:
- get rid of this interface
- get rid of `timingHistogramVecElement` completely 

Which will simplify this code.

Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go]: done

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+	if err != nil {
+		return noopRatioed{}, err
+	}
+	underMember.Set(initialNumerator / initialDenominator)
+	return &timingRatioHistogramInner{
+			getGaugeOfRatio: func() Gauge { return underMember },
+			numerator:       initialNumerator,
+			denominator:     initialDenominator,
+		},
+		nil
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:185]: nit: move to the previous line:
```
...
}, nil
```

**Review by wojtek-t — COMMENTED** [review]
I just have two small nits - other than that LGTM.

But I would like to see an LGTM from @dgrisonnet on the actual metrics changes.
@dgrisonnet - can you please take a look?

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+// NewForLabelValuesChecked will return an error if this vec is not hidden and not yet registered
+// or there is a syntactic problem with the labelValues.
+func (v *TimingRatioHistogramVec) NewForLabelValuesChecked(initialNumerator, initialDenominator float64, labelValues []string) (RatioedGauge, error) {
+	underMember, err := v.delegate.WithLabelValuesChecked(labelValues...)
+	if err != nil {
+		return noopRatioed{}, err
+	}
+	underMember.Set(initialNumerator / initialDenominator)
+	return &timingRatioHistogramInner{
+			getGaugeOfRatio: func() Gauge { return underMember },
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:181]: github makes it hard (but not impossible) to select the initial tab on each line.  Line 180 starts with one tab, line 181 starts with three.  This is what `gofmt` produces.

[the remark this answers is not in the document]
Comment (wojtek-t) [reply staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:181]: But it's even inconsistent with line 205-209...

**Comment by wojtek-t** [comment]
LGTM, but I want an LGTM from sig-instrumention folks before approving.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 17]
```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:2]: Wrong year.

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+// `*TimingRatioHistogram` implements both Registerable and RatioedGauge.
+type TimingRatioHistogram struct {
+	// The implementation is layered on TimingHistogram,
+	// adding the division by an occasionally adjusted denominator.
+
+	// Registerable is the registerable aspect.
+	// That is the registerable aspect of the underlying TimingHistogram.
+	compbasemetrics.Registerable
+
+	// timingRatioHistogramInner implements the RatioedObsserver aspect.
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go:41]: s/RatioedObsserver/RatioedObserver/

File: staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram_test.go [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+const (
+	ddtRangeCentiPeriods  = 300
+	ddtOffsetCentiPeriods = 50
+	numIterations         = 100
+)
+
+var errMetricNotFound = errors.New("not found")
+
+func TestTimingRatioHistogramVecElementSimple(t *testing.T) {
+	samplesHistName := "tro_test"
```
Comment (logicalhan) [inline staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram_test.go:40]: This can basically be a constant, the way it is used in the tests.

**Review by logicalhan — COMMENTED** [review]
/lgtm
(from the sig-instrumentation side of things)

I'm okay with the metrics changes here, since they are alpha metrics.

**Comment by MikeSpreitzer** [comment]
The latest force-push makes the changes suggested by @logicalhan .

**Comment by MikeSpreitzer** [comment]
And the latest force-push changes the timing ratio histogram test funcs to each use a distinct name for the test histogram, for additional log clarity and possibly correctness in parallel testing.

**Comment by wojtek-t** [comment]
Thanks @logicalhan  - lgtming with your LGTM from the sig-instrumentation POV.

/lgtm
/approve

/retest
